# <center> Data Collection and Cleaning

**Overview** - The notebook serves the purpose of importing the data, converting the needed features to the desired datatypes, handling null values etc. <br>
**Result** - The cleaned data is the exported to be used further by the visualisation and prediction notebook<br>
**Process** - <br>
 - [Importing the Libraries](#Installing-Libraries)
 - [Importing the Data](#Importing-the-Data)
 - [Examining the Data](#Examining-the-Data)
 - [Removing null values and unwanted columns](#Removing-null-values-and-unwanted-columns)


## Installing Libraries


We will install the needfull libraries over here using the ```!pip install``` command.<br>
Libraries used are:<br>
- Pandas and Numpy : For data handling
- Seaborn and Matplotlib : For Visualisation
- Projects : To retrieve and add data to the project
`

This followed by importing the libraries for further use


!pip install pandas
!pip install numpy
!pip install seaborn
!pip install matplotlib

In [565]:
project_token = 'p-3f704f3cdf69a8a61cd7aa18710a153652a855a1'

In [566]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
from project_lib import Project
project = Project( project_access_token=project_token)
pc = project.project_context

## Importing the Data


The data is imported to the notebbok from the assets section of the project<br>
The code over here is imported directly from the watson data functionality

In [567]:
# The code was removed by Watson Studio for sharing.

We will install the needfull libraries over here using the ```.read_csv``` command of Pandas<br>
The imported csv is stored as a dtaframe with the name ```data```.

In [568]:
data = pd.read_csv(body)

Viewing the head of the data

In [569]:
data.head()

Source.Name web-scraper-order  \
0  cardekho_extract(0-2000).csv   1611917819-1662   
1  cardekho_extract(0-2000).csv   1611918361-1902   
2  cardekho_extract(0-2000).csv   1611917012-1306   
3  cardekho_extract(0-2000).csv   1611917695-1607   
4  cardekho_extract(0-2000).csv    1611914861-367   

                               web-scraper-start-url  \
0  https://www.cardekho.com/used-car-details/used...   
1  https://www.cardekho.com/used-car-details/used...   
2  https://www.cardekho.com/used-car-details/used...   
3  https://www.cardekho.com/used-car-details/used...   
4  https://www.cardekho.com/used-car-details/used...   

                                        full_name selling_price  \
0                                 Maruti Alto Std     1.2 Lakh*   
1                          Hyundai Grand i10 Asta     5.5 Lakh*   
2                                Hyundai i20 Asta    2.15 Lakh*   
3                   Maruti Alto K10 2010-2014 VXI    2.26 Lakh*   
4  Ford Ecosport 2015-2021 1.5 TDCi Titanium BSIV     5.7 Lakh*   

                                        new-price    year seller_type  \
0                                             NaN  2012.0  Individual   
1    New Car (On-Road Price) : Rs.7.11-7.48 Lakh*  2016.0  Individual   
2                                             NaN  2010.0  Individual   
3                                             NaN  2012.0  Individual   
4  New Car (On-Road Price) : Rs.10.14-13.79 Lakh*  2015.0      Dealer   

      km_driven   owner_type fuel_type transmission_type            mileage  \
0  1,20,000 kms  First Owner    Petrol            Manual   Mileage19.7 kmpl   
1    20,000 kms  First Owner    Petrol            Manual   Mileage18.9 kmpl   
2    60,000 kms  First Owner    Petrol            Manual   Mileage17.0 kmpl   
3    37,000 kms  First Owner    Petrol            Manual  Mileage20.92 kmpl   
4    30,000 kms  First Owner    Diesel            Manual  Mileage22.77 kmpl   

          engine           max_power   seats  
0   Engine796 CC   Max Power46.3 bhp  Seats5  
1  Engine1197 CC     Max Power82 bhp  Seats5  
2  Engine1197 CC     Max Power80 bhp  Seats5  
3   Engine998 CC   Max Power67.1 bhp  Seats5  
4  Engine1498 CC  Max Power98.59 bhp  Seats5

## Examining the Data

Now since the data is imported we will see the composition and various data types and look for null values etc.

Looking at the data shape

In [570]:
data.shape

(20026, 16)

So now we know that there are **20026** observations and **16** columns

Looking at data types of each column

In [571]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20026 entries, 0 to 20025
Data columns (total 16 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Source.Name            20026 non-null  object 
 1   web-scraper-order      20026 non-null  object 
 2   web-scraper-start-url  20026 non-null  object 
 3   full_name              19980 non-null  object 
 4   selling_price          19980 non-null  object 
 5   new-price              9566 non-null   object 
 6   year                   19980 non-null  float64
 7   seller_type            19980 non-null  object 
 8   km_driven              19980 non-null  object 
 9   owner_type             19980 non-null  object 
 10  fuel_type              19980 non-null  object 
 11  transmission_type      19980 non-null  object 
 12  mileage                19980 non-null  object 
 13  engine                 19921 non-null  object 
 14  max_power              19921 non-null  object 
 15  se

Null Values (absolute and percentage) by column
<br> This shows how many columns have a very high composition of null values rendering them useless

In [572]:
Column,NullValues,NullValuespercentage = [],[],[]
for i in data.columns:
    Column.append(i)
    NullValues.append(data[i].isnull().sum())
    NullValuespercentage.append(round(data[i].isnull().mean(),3))
display(pd.DataFrame({'Column':Column,'Number of Null Values':NullValues,'% of null values':NullValuespercentage}).sort_values(by='% of null values',ascending=False).set_index('Column'))

Number of Null Values  % of null values
Column                                                        
new-price                              10460             0.522
seats                                    173             0.009
engine                                   105             0.005
max_power                                105             0.005
full_name                                 46             0.002
selling_price                             46             0.002
year                                      46             0.002
seller_type                               46             0.002
km_driven                                 46             0.002
owner_type                                46             0.002
fuel_type                                 46             0.002
transmission_type                         46             0.002
mileage                                   46             0.002
Source.Name                                0             0.000
web-scraper-order                          0             0.000
web-scraper-start-url                      0             0.000

## Removing null values and unwanted columns

The _new-price_ column contains more null values than can be handled so this column will be deleted.<br>
So will the columns _web-scraper-start-url, web-scraper-order_ and _Source.Name_ as they do not cnovey any meaningful information iand merely there as external refernces

In [573]:
to_be_removed = ['web-scraper-start-url','Source.Name','web-scraper-order','new-price']
data = data.drop(to_be_removed,axis=1)

Rechecking for null values

In [574]:
Column,NullValues,NullValuespercentage = [],[],[]
for i in data.columns:
    Column.append(i)
    NullValues.append(data[i].isnull().sum())
    NullValuespercentage.append(round(data[i].isnull().mean(),3))
display(pd.DataFrame({'Column':Column,'Number of Null Values':NullValues,'% of null values':NullValuespercentage}).sort_values(by='% of null values',ascending=False).set_index('Column'))

Number of Null Values  % of null values
Column                                                    
seats                                173             0.009
engine                               105             0.005
max_power                            105             0.005
full_name                             46             0.002
selling_price                         46             0.002
year                                  46             0.002
seller_type                           46             0.002
km_driven                             46             0.002
owner_type                            46             0.002
fuel_type                             46             0.002
transmission_type                     46             0.002
mileage                               46             0.002

Since the highest number of null values is still **less than 0.10%**, we can simply delete those entries

In [575]:
data = data.dropna()

Now with null values removed, we can proceed to working on the features

## Data Correction

In [576]:
data.head()

full_name selling_price    year  \
0                                 Maruti Alto Std     1.2 Lakh*  2012.0   
1                          Hyundai Grand i10 Asta     5.5 Lakh*  2016.0   
2                                Hyundai i20 Asta    2.15 Lakh*  2010.0   
3                   Maruti Alto K10 2010-2014 VXI    2.26 Lakh*  2012.0   
4  Ford Ecosport 2015-2021 1.5 TDCi Titanium BSIV     5.7 Lakh*  2015.0   

  seller_type     km_driven   owner_type fuel_type transmission_type  \
0  Individual  1,20,000 kms  First Owner    Petrol            Manual   
1  Individual    20,000 kms  First Owner    Petrol            Manual   
2  Individual    60,000 kms  First Owner    Petrol            Manual   
3  Individual    37,000 kms  First Owner    Petrol            Manual   
4      Dealer    30,000 kms  First Owner    Diesel            Manual   

             mileage         engine           max_power   seats  
0   Mileage19.7 kmpl   Engine796 CC   Max Power46.3 bhp  Seats5  
1   Mileage18.9 kmpl  Engine1197 CC     Max Power82 bhp  Seats5  
2   Mileage17.0 kmpl  Engine1197 CC     Max Power80 bhp  Seats5  
3  Mileage20.92 kmpl   Engine998 CC   Max Power67.1 bhp  Seats5  
4  Mileage22.77 kmpl  Engine1498 CC  Max Power98.59 bhp  Seats5

The _selling_price_ should be numeric in order to draw work easily on it, hence the next step is to remove the unwanted characters and convert the column to a float type

In [ ]:
selling_price = []
for i in data['selling_price'].str.upper():
    if "LAND ROVER" in i:
        temp = i.split(',')[0]
        temp = int(temp)/100
        selling_price.append(temp)
    else:
        selling_price.append(float(i))
data['selling_price'] = selling_price
data.rename(columns = {'selling_price':'selling_price (in lakhs)'}, inplace = True)

Firstly, the _full_name_ should be split into _model, maker_ and _variant_ ro help with visualsiation later on

In [577]:
maker = data["full_name"].str.upper().str.split(" ", n = 2, expand = True)[0]
data['maker']=maker
model = data["full_name"].str.upper().str.split(" ", n = 2, expand = True)[1]      
data['model']=model
variant = data["full_name"].str.upper().str.split(" ", n = 2, expand = True)[2]
data['variant']=variant
data = data.drop('full_name',axis=1)
for i in data.index:
    if data['variant'][i] == None:
        data['variant'][i] = data['model'][i]
#data.head()

<ipython-input-577-6f3d4454fbc7>:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['variant'][i] = data['model'][i]


Now, converting the _year_ column to the _age_ column will help as it reduces the scale of the numbers and is easier to interpret

In [578]:
data['year']=2021 - data['year']
data.rename(columns = {'year':'age'}, inplace = True)
#data.head()

The _km_driven_ should be numeric in order to draw work easily on it, hence the next step is to remove the unwanted characters and convert the column to a float type

In [580]:
data['km_driven'] = data["km_driven"].str.split(" kms", n = 1, expand = True)[0].str.replace(',', '').astype(int)/1000
data.rename(columns = {'km_driven':'km_driven (in thousands)'}, inplace = True)
#data.head()

The _mileage_ should be numeric in order to draw work easily on it, hence the next step is to remove the unwanted characters and convert the column to a float type

In [581]:
data['mileage'] = data["mileage"].str.split("kmpl", n = 1, expand = True)[0].str.split("Mileage", n = 1, expand = True)[1]
data['mileage'] = data["mileage"].str.split(" km/kg", n = 1, expand = True)[0].str.split(" km/hr", n = 1, expand = True)[0]

data['mileage'] = data["mileage"].astype(float)
#data.head() km/hr

The _engine_ should be numeric in order to draw work easily on it, hence the next step is to remove the unwanted characters and convert the column to a integer type

In [582]:
data['engine'] = data["engine"].str.split(" CC", n = 1, expand = True)[0].str.split("Engine", n = 1, expand = True)[1]
data = data[data['engine']!='']
data['engine'] = data['engine'].astype(int)
#The above line is done to remove any null entries which remain after strings have been removed
data.rename(columns = {'engine':'engine_size (in cc)'}, inplace = True)

The _max_power_ should be numeric in order to draw work easily on it, hence the next step is to remove the unwanted characters and convert the column to a float type

In [583]:
data['max_power'] = data["max_power"].str.split(" bhp", n = 1, expand = True)[0].str.split("Power", n = 1, expand = True)[1]
data=data[data['max_power']!='null']
data=data[data['max_power']!='']
#The above line is done to remove any null entries which remain after strings have been removed
data['max_power'] = data['max_power'].astype(float)
data.rename(columns = {'max_power':'max_power (in bhp)'}, inplace = True)

The _seats_ should be numeric in order to draw work easily on it, hence the next step is to remove the unwanted characters and convert the column to a integer type

In [584]:
data['seats'] = data["seats"].str.split("Seats", n = 1, expand = True)[1]
data=data[data['seats']!='N/A']
data['seats'] = data['seats'].astype(int)

post resetting the index, we can see the finished dataset and compare it with the one imported

In [585]:
data = data.reset_index().drop('index',axis=1)

In [586]:
data.head()

selling_price (in lakhs)   age seller_type  km_driven (in thousands)  \
0                      1.20   9.0  Individual                     120.0   
1                      5.50   5.0  Individual                      20.0   
2                      2.15  11.0  Individual                      60.0   
3                      2.26   9.0  Individual                      37.0   
4                      5.70   6.0      Dealer                      30.0   

    owner_type fuel_type transmission_type  mileage  engine_size (in cc)  \
0  First Owner    Petrol            Manual    19.70                  796   
1  First Owner    Petrol            Manual    18.90                 1197   
2  First Owner    Petrol            Manual    17.00                 1197   
3  First Owner    Petrol            Manual    20.92                  998   
4  First Owner    Diesel            Manual    22.77                 1498   

   max_power (in bhp)  seats    maker     model  \
0               46.30      5   MARUTI      ALTO   
1               82.00      5  HYUNDAI     GRAND   
2               80.00      5  HYUNDAI       I20   
3               67.10      5   MARUTI      ALTO   
4               98.59      5     FORD  ECOSPORT   

                            variant  
0                               STD  
1                          I10 ASTA  
2                              ASTA  
3                 K10 2010-2014 VXI  
4  2015-2021 1.5 TDCI TITANIUM BSIV

In [587]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19631 entries, 0 to 19630
Data columns (total 14 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   selling_price (in lakhs)  19631 non-null  float64
 1   age                       19631 non-null  float64
 2   seller_type               19631 non-null  object 
 3   km_driven (in thousands)  19631 non-null  float64
 4   owner_type                19631 non-null  object 
 5   fuel_type                 19631 non-null  object 
 6   transmission_type         19631 non-null  object 
 7   mileage                   19631 non-null  float64
 8   engine_size (in cc)       19631 non-null  int64  
 9   max_power (in bhp)        19631 non-null  float64
 10  seats                     19631 non-null  int64  
 11  maker                     19631 non-null  object 
 12  model                     19631 non-null  object 
 13  variant                   19631 non-null  object 
dtypes: flo

The cleaned data is just **395** entries smaller than the original dataset had **20026** and gives us a much more easily useable file

## Export

In [590]:
project.save_data(data=data.to_csv(index=False),file_name='car_data_cleaned.csv',overwrite=True)

{'file_name': 'car_data_cleaned.csv',
 'message': 'File saved to project storage.',
 'bucket_name': 'carresaleanalysisandprediction-donotdelete-pr-emys7cc9x01d89',
 'asset_id': 'b2d93b21-1958-4dca-990d-477adfd5d044'}

Now the file will be xported to the project as a new csv

## Notes

### Author 
Created by **Bhavesh Jain** <br>
Github link - https://github.com/bhaveshjain1612 <br>
Linkedin Link - https://www.linkedin.com/in/bhaveshjain1612/ <br>
Kaggle link - https://www.kaggle.com/bhaveshjain1612 <br>

### Changelog


| Date      | Name | Change     |
| :---        |    :----:   |          ---: |
| 27/09/2021      | Bhavesh Jain       | Created Notebook  |
| 28/09/2021   | Bhavesh Jain        | Added Added Notes Section      |